keps/sig-api-machinery/3920-finish-feature-flags/README.md

```diff
@@ -0,0 +1,951 @@
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Kubernetes uses feature gates for a wide variety of features. Unfortunately,
+there are many problems and missing ablities with the current system (see
+below). This KEP solves these problems. Since it is wide reaching, we break
+implementation into several stages:
+
+Stage 1:
+* Add a Feature API
+* Components report their use of the feature (i.e., how it is set on their
+  command line)
+* Add a control-plane enablement/disablement flow (example: clearing fields in
+  etcd)
+
+Stage 2:
+* Add a cluster upgrade / downgrade flow beside the enablement flow (example:
+  clearing a collection where the alpha and beta schemas are incompatible)
+* For selected features, get the desired state from the API (it is set on
+  apiserver's command line)
+
+Stage 3:
+* If opted in on apiserver's command line, a feature may be set dynamically in
+  the API.
+
+This functionality needs both stages 1 and 2 to proceed to beta or GA. Stage 3 is
+optional but will be easy.
+
+## Motivation
+
+A list of problems:
+* It's mostly unsafe to turn a feature on, use it, and turn it back off. This is
+  because state stored while the feature was on is not cleaned up. So instead of
+  going back into a pristine state, the cluster goes into a third state,
+  on-with-dangling-references. This third state is tested less well than either
+  the on or off states.
+* Because of this, it's risky to try alpha features with important clusters.
+* It's not very useful or convenient to make disposable clusters to try alpha
+  features.
+* Because of that, it's extremely hard to get robust feedback about features
+  while they are in alpha and easy to change.
+* There is no place in the code for devs to put turn on / turn off code, even if
+  they wanted to.
+* Likewise there is no place to put upgrade or downgrade code. Many efforts have
+  used built-in resources instead of CRDs simply because there is no place to
+  put CRD install code.
+* It is difficult to know if a feature has been turned on for every relevant
+  binary. Yes, there are metrics, but you have to know which binaries are
+  relevant for what features and your particular cluster's metrics collection
+  system.
+* It is especially difficult to know if use of a feature at some point makes
+  cluster upgrade unsafe now; for example, an alpha object being written while
+  the feature was briefly on which is incompatable with the default-on beta.
+* It is not easy to test the feature on -> use -> feature off lifecycle.
+
+### Goals
+
+Primary goals:
+* Minimize codebase churn, existing feature authors do little-to-no work
+* Expose the feature states in use in the cluster
+* Make it possible to get the cluster back into the "off" state
+* Expose the "off but dangling references / uses" state
+* Make it easy to add enablement and disablement code
+* Make it easy to add upgrade and downgrade code
```

Comment:
Would this (eventually) obviate the need for storage version migrator?

Reply:
Probably not, but this would provide a convenient place to trigger it, or wait for it to complete.